sdk/component/target_test.go

```diff
@@ -0,0 +1,133 @@
+package component_test
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
+
+	"github.com/smithy-security/smithy/sdk/component"
+	"github.com/smithy-security/smithy/sdk/component/internal/mocks"
+)
+
+func runTargetHelper(t *testing.T, ctx context.Context, target component.Target) error {
+	t.Helper()
+
+	return component.RunTarget(
+		ctx,
+		target,
+		component.RunnerWithLogger(component.NewNoopLogger()),
+		component.RunnerWithComponentName("sample-target"),
+	)
+}
+
+func TestRunTarget(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+
+	var (
+		ctrl = gomock.NewController(t)
+	)
+
+	t.Run("it should run a target correctly", func(t *testing.T) {
+		mockTarget := mocks.NewMockTarget(ctrl)
+
+		mockTarget.
+			EXPECT().
+			Prepare(gomock.Any()).
+			Return(nil)
+		mockTarget.
+			EXPECT().
+			Close(gomock.Any()).
+			Return(nil)
+
+		require.NoError(t, runTargetHelper(t, ctx, mockTarget))
+	})
+
+	t.Run("it should return early when the context is cancelled", func(t *testing.T) {
+		ctx, cancel = context.WithTimeout(ctx, 100*time.Millisecond)
+		defer cancel()
+
+		mockTarget := mocks.NewMockTarget(ctrl)
+
+		mockTarget.
+			EXPECT().
+			Prepare(gomock.Any()).
+			DoAndReturn(func(ctx context.Context) error {
+				<-ctx.Done()
+				return nil
+			})
+		mockTarget.
+			EXPECT().
+			Close(gomock.Any()).
+			Return(nil)
+
+		require.NoError(t, runTargetHelper(t, ctx, mockTarget))
+	})
+
+	t.Run("it should return early when prepare errors", func(t *testing.T) {
+		var (
+			errPrepare = errors.New("prepare-is-sad")
+			mockTarget = mocks.NewMockTarget(ctrl)
+		)
+
+		mockTarget.
+			EXPECT().
+			Prepare(gomock.Any()).
+			Return(errPrepare)
+		mockTarget.
+			EXPECT().
+			Close(gomock.Any()).
+			Return(nil)
+
+		err := runTargetHelper(t, ctx, mockTarget)
+		require.Error(t, err)
+		assert.ErrorIs(t, err, errPrepare)
```

Comment:
why use an assert here instead of require? leaving this comment here, but it's present in various places

Reply:
Nothing bad in using an assert here as the next steps in the test case don't depend on it's success - also there are no steps next!

In general I find using `require` great for when we MUST assert that something is valid before moving on - for example, checking that a slice has 2 elements before doing assertions on those two elements

`assert` on the other end, doesn't stop the test if it fails and it's great to be used when we have multiple assertions in a block that wouldn't cause other issues in the test case if fail. The feedback loop of development using assert for failing test cases is faster than `require`.

If you try this example you'll get what I mean!

```go
func TestAssert(t *testing.T) {
	users := []user{
		{name: "Andrea", email: "andrea@smithy.security.com"},
		{name: "Pavlos", email: "pavlos@smithy.security.com"},
	}

	require.Len(t, users, 2)
	assert.Equal(t, "Diana", users[0])
	assert.Equal(t, "Spyros", users[1])
}

func TestRequire(t *testing.T) {
	users := []user{
		{name: "Andrea", email: "andrea@smithy.security.com"},
		{name: "Pavlos", email: "pavlos@smithy.security.com"},
	}

	require.Len(t, users, 2)
	require.Equal(t, "Diana", users[0])
	require.Equal(t, "Spyros", users[1])
}
```